arch/xtensa/src/esp32/esp32_psram.c

```diff
@@ -0,0 +1,1713 @@
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: esp32_get_vddsdio_config
+ *
+ * Description:
+ *   Get current VDDSDIO configuration.
+ *
+ * Input Parameters:
+ *   Incoming parameter address of VDDSDIO configuration to be saved
+ *
+ * Returned Value:
+ *   Zero (OK) is returned on success.
+ *
+ ****************************************************************************/
+
+static int IRAM_ATTR esp32_get_vddsdio_config(
+                     struct rtc_vddsdio_config_s *config)
+{
+  struct rtc_vddsdio_config_s *result = config;
+  uint32_t efuse_reg;
+  uint32_t strap_reg;
+  uint32_t sdio_conf_reg = getreg32(RTC_CNTL_SDIO_CONF_REG);
+
+  result->drefh = (sdio_conf_reg & RTC_CNTL_DREFH_SDIO_M)
+                                 >> RTC_CNTL_DREFH_SDIO_S;
+  result->drefm = (sdio_conf_reg & RTC_CNTL_DREFM_SDIO_M)
+                                 >> RTC_CNTL_DREFM_SDIO_S;
+  result->drefl = (sdio_conf_reg & RTC_CNTL_DREFL_SDIO_M)
+                                 >> RTC_CNTL_DREFL_SDIO_S;
+
+  if (sdio_conf_reg & RTC_CNTL_SDIO_FORCE)
+    {
+      /* Get configuration from RTC */
+
+      result->force = 1;
+      result->enable = (sdio_conf_reg & RTC_CNTL_XPD_SDIO_REG_M)
+                                      >> RTC_CNTL_XPD_SDIO_REG_S;
+      result->tieh = (sdio_conf_reg & RTC_CNTL_SDIO_TIEH_M)
+                                      >> RTC_CNTL_SDIO_TIEH_S;
+
+      return OK;
+    }
+
+  efuse_reg = getreg32(EFUSE_BLK0_RDATA4_REG);
+
+  if (efuse_reg & EFUSE_RD_SDIO_FORCE)
+    {
+      /* Get configuration from EFUSE */
+
+      result->force = 0;
+      result->enable = (efuse_reg & EFUSE_RD_XPD_SDIO_REG_M)
+                                  >> EFUSE_RD_XPD_SDIO_REG_S;
+      result->tieh = (efuse_reg & EFUSE_RD_SDIO_TIEH_M)
+                                >> EFUSE_RD_SDIO_TIEH_S;
+
+      if (REG_GET_FIELD(EFUSE_BLK0_RDATA3_REG,
+              EFUSE_RD_BLK3_PART_RESERVE) == 0)
+        {
+          result->drefh = (efuse_reg & EFUSE_RD_SDIO_DREFH_M)
+                                     >> EFUSE_RD_SDIO_DREFH_S;
+          result->drefm = (efuse_reg & EFUSE_RD_SDIO_DREFM_M)
+                                     >> EFUSE_RD_SDIO_DREFM_S;
+          result->drefl = (efuse_reg & EFUSE_RD_SDIO_DREFL_M)
+                                     >> EFUSE_RD_SDIO_DREFL_S;
+        }
+
+      return OK;
+    }
+
+  /* Otherwise, VDD_SDIO is controlled by bootstrapping pin */
+
+  strap_reg = getreg32(GPIO_STRAP_REG);
+  result->force = 0;
+  result->tieh = (strap_reg & BIT(5)) ? RTC_VDDSDIO_TIEH_1_8V
+                                      : RTC_VDDSDIO_TIEH_3_3V;
+  result->enable = 1;
+
+  return OK;
+}
+
+static void IRAM_ATTR
+psram_cache_init(psram_cache_mode_t psram_cache_mode,
+                 psram_vaddr_mode_t vaddrmode);
+
```

Comment:
Please move this to Private Function Prototypes

Reply:
Done!